lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,64 +12,72 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
-  ExprVector expTypeStack;
-
-  bool isConst (Expr exp) {
-   return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  ExprMap exprMap;
+
+  bool isBool(Expr exp) {
+    if (isOpX<TRUE>(exp) || isOpX<FALSE>(exp))
+      return true;
+    else if (bind::isBoolVar(exp) || bind::isBoolConst(exp))
+      return true;
+    else if (isOpX<AND>(exp) || isOpX<OR>(exp) || isOpX<XOR>(exp) ||
+             isOpX<IMPL>(exp) || isOpX<ITE>(exp) || isOpX<IFF>(exp) ||
+             isOpX<NEG>(exp))
+      return true;
+
+    return false;
   }
 
-  bool isVar (Expr exp) {
-   return bind::isBoolVar(exp) || bind::isIntVar(exp);
-  }
+  Expr inferType(Expr exp /*,TypeChecker & ty*/) {
+    if (isBool(exp))
+      return sort::boolTy(exp->efac());
+    else if (bind::isIntVar(exp) || bind::isIntConst(exp))
+      return sort::intTy(exp->efac());
 
-  bool isValue (Expr exp) {
-    return isOpX<TRUE>(exp) || isOpX<FALSE>(exp);
+    return sort::anyTy(exp->efac());
   }
 
-  //returns true if children are of correct type
-  bool checkChildren (Expr exp) {
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp);
+  }
 
-    if (isOpX<ITE>(exp)) {
-      //ite (a,b,c) : a is bool, b and c same type
-      Expr c = expTypeStack.back(); 
-      expTypeStack.pop_back();
+  bool isVar(Expr exp) { return bind::isBoolVar(exp) || bind::isIntVar(exp); }
 
-      Expr b = expTypeStack.back(); 
-      expTypeStack.pop_back();
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
 
-      Expr a = expTypeStack.back(); 
-      expTypeStack.pop_back();
+  // returns true if children are of correct type
+  bool checkChildren(Expr exp) {
 
-      bool properTypes = (a == sort::boolTy (exp->efac())) && (b == c);
-      return properTypes;
+    if (isOpX<ITE>(exp)) {
+      // ite (a,b,c) : a is bool, b and c same type
+      bool aIsBool = exprMap[exp->arg(0)] == sort::boolTy(exp->efac());
```

Comment:
wrap calls to `exprMap` with a getter function. Cache `boolTy`, don't use `operator[]`, use `.at()` instead.

Reply:
use C++ algorithms to check that all arguments are of the same type: https://en.cppreference.com/w/cpp/algorithm/all_any_none_of